packages/kbn-content-management-utils/types.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  GetIn,
+  GetResult,
+  CreateIn,
+  CreateResult,
+  SearchIn,
+  SearchResult,
+  UpdateIn,
+  UpdateResult,
+  DeleteIn,
+  DeleteResult,
+} from '@kbn/content-management-plugin/common';
+
+export interface Reference {
+  type: string;
+  id: string;
+  name: string;
+}
+
+export interface CreateOptions {
```

Comment:
I wonder if it is a good idea to lock the same `Options` interface for all content type. The idea was to allow each content to pass different options to the storage layer and type those interface + add a schema for them.

Maybe you are right that all call to create/update/delete/search will have the exact same options object but from where I am standing I can't make such a call. WDYT?

Reply:
Let me see if I understand correctly.

I _only_ added the `references` prop to create because that's the only `SavedObjectCreateOptions` that **map** uses. They are more fields in that interface but "map" does not use them so I limited the interface. 

Here you are making this decision generic for all SO CM types..

Now if "lens" needs another option to be passed we will allow all other types to provide that other option, even though it is not related to they need and it does not do anything.

It seems to me that we are back in tying different content together and it might be hard to make evolve interfaces with time and version them separately.

* How will we know if we need to up/down transform an option object?
* How will you provide a different validation schema if "lens" need to pass this options object: `{ myLensOptionField: true, references: ['123'] }`. Will now all CM types have to up/down transform this specific `myLensOptionField`, what will be the code line documentation for that "specific to lens" option?

Reply:
>Here you are making this decision generic for all SO CM types..

If you'd like, we could provide functionality to extend the Option types. 

---

>How will we know if we need to up/down transform an option object?

I've ignored that since we don't yet know how the Saved Object versioning will work, but perhaps it can be considered separately.

Reply:
We should not have any well defined "options" across all CM types. If there is some field that would be useful for all CM types, it should not be an option, but built into the CM CRUD API directly.

Options should ideally never be used, they are an escape hatch, that can be utilised when some type has some very extra behavior that does not fill well into the starndartized CM CURD API.

Having said that, I don't think any database related concern should be standartized in the CM layer. For example, *references* are a database concern (some database have them, some don't, some content types use them, some don't), the CM system should not know anything about the Saved Object references.

Also, with regards to TypeScript types: in CM plugin and packages there should be no types that deal with SavedObjects, i.e. all names that contain `*SO*`, `savedObject*`, etc.. do not belong to CM code. Those should be handled on the content provider level.

For example, if AX group things it is useful for their content types to standartize some "options" or SO functionality, you are free to do that, but that code should live outside of CM.

Reply:
@vadimkibana 

>Options should ideally never be used, they are an escape hatch, that can be utilised when some type has some very extra behavior that does not fill well into the starndartized CM CURD API.

Does that mean they shouldn't be used as they are for Maps?

---

>Having said that, I don't think any database related concern should be standartized in the CM layer

I'm okay with that and understand it in abstract. That said, we need a Saved Objects layer! I'm getting a bit frustrated here to the point where it feels like there's been some significant miscommunication. I understand that there's value in providing abstraction that allows for different methods of persisting content management items BUT as best I can tell, the majority of cases are going to be Saved Objects, at least for the foreseeable future. 

Failing to provide an easy to implement wrapper for Saved Objects feels like a choice just to make a point. It doesn't benefit Content Management api consumers working with Saved Objects, at least not that I can see. 

Applying what I learned from the Maps implementation to the Data Views implementation was a big copy and paste project. Seems like that should be avoided.

Reply:
> If you'd like, we could provide functionality to extend the Option types.

That was not my point. We don't want to extend it. We want it to be unique to each content.
Some content type might not have any options at all.

```ts
// Valid create lens call (no options allowed)
await cmClient.create('lens', { ...lensData });

// Invalid create lens call
// This should **not** be allowed if "lens" does not uses "references"
await cmClient.create('lens', { ...lensData }, { references: ['123'] });
```

Reply:
>We want it to be unique to each content.

I understand what you're saying, but as someone considering how I'd apply the Content Management API to a number of saved objects, it doesn't make sense. 

Why don't we want to reuse code in this circumstance?

Reply:
> Applying what I learned from the Maps implementation to the Data Views implementation was a big copy and paste project.

What I'm saying if you feel like you want to standartize "options" across your types, that is all good. But that standartized code should not live in CM, it could be some "AX storage helpers" code.

Reply:
@vadimkibana 

This is in a new `content-management-utils` package - is that good?

Reply:
> This is in a new `content-management-utils` package - is that good?

Yes, if that package is not expected to be owned by SharedUx.

Reply:
@vadimkibana @sebelga 

I think we should find some time to talk. Its unclear why some decisions have been made regarding the CM api and I'd like clarity before I continue work on this.